tbd/handler/handler.go

```diff
@@ -0,0 +1,72 @@
+package handler
+
+import (
+	"fmt"
+	"net/http"
+)
+
+type Logger interface {
+	Request(*http.Request, error)
+}
+
+type Handler func(w http.ResponseWriter, r *http.Request) error
+
+type ErrorFunc func(w http.ResponseWriter, r *http.Request, statusCode int, err error)
+
+func New(redirectBaseURL, authURL string, errorFunc ErrorFunc) func(Handler) http.Handler {
+	return func(next Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if err := next(w, r); err != nil {
+				statusCode := http.StatusInternalServerError
+
+				switch v := err.(type) {
+				case RedirectError:
+					http.Redirect(w, r, redirectBaseURL+v.RedirectTo(), http.StatusFound)
+					return
+				case StatusError:
+					if v.StatusCode() == http.StatusUnauthorized {
+						http.Redirect(w, r, authURL, http.StatusFound)
+						return
+					}
+
+					if v.StatusCode() == http.StatusForbidden || v.StatusCode() == http.StatusNotFound {
+						statusCode = v.StatusCode()
+					}
+				}
+
+				w.WriteHeader(statusCode)
+				errorFunc(w, r, statusCode, err)
+			}
+		})
+	}
+}
+
+type RedirectError interface {
+	RedirectTo() string
+}
+
+type Redirect string
+
+func (e Redirect) Error() string {
+	return "redirect to " + string(e)
+}
+
+func (e Redirect) RedirectTo() string {
+	return string(e)
+}
+
+type StatusError interface {
```

Comment:
My reasoning behind this is that it allows the `sirius` package (or something else) to have an error that when handled logs you out. But another thought I had was to split that to it's own interface.

```
type UnauthorizedError interface {
  IsUnauthorized() bool
}
```

Then we could add:

```
func (e Status) IsUnauthorized() bool {
  return int(e) == http.StatusUnauthorized
}

// in internal/sirius/client.go
func (e ClientError) IsUnauthorized() bool {
  return e == ErrUnauthorized
}
```

:shrug: